Add scrollbars to app

tunjid opened the pull request:
This PR adds scrollbar support to lazy containers in the app.

The scrollbar Composable is a stand alone component with the following definition:

* A ScrollbarState: Defines the thumb size and position on the track
* A thumb Composable: Allows users to define the thumb according to their design system
* A minimum thumb size: For accessible touch targets
* An interaction source: For styling depending on interaction state
* Thumb interaction callback: To react to thumb interactions from the user

Please run in release mode to test.

Phones:
[scrollbars.webm](https://github.com/android/nowinandroid/assets/8147224/7d49b9eb-8cbb-406d-81aa-ec7909a7f5b2)
![scroll](https://github.com/android/nowinandroid/assets/8147224/3fc048c5-b42b-4b3a-95e7-f2649f7c390c)


Tablets:
[tabs.webm](https://github.com/android/nowinandroid/assets/8147224/587661b0-09cf-42f8-8f07-e5917e249e0e)
![t](https://github.com/android/nowinandroid/assets/8147224/c74cb7ab-b49d-46da-9517-9f457a7d4140)

Chromebooks
![desktop](https://github.com/android/nowinandroid/assets/8147224/7a953ede-3de0-4baa-abd3-b155bcdec037)

The PR:

* Offers fast scroll where beneficial. Users can also long press the scroll bar track to fast scroll.
* Offers a simple indicative scrollbar where appropriate
* Offers it for both lazy grids ands lazy lists with the algorithm generally applicable to lazy staggered grids and simple simple scroll states.

The scrollbars are unobtrusive as they disappear when the list is dormant. It has 3 states:

* Active: The user is actively pressing or dragging the scrollbar in a fast scroll, or the list is being scrolled
* Inactive: Cooldown period after being dragged
* Dormant: Inert static list

mmoczkowski commented on core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/component/scrollbar/AppScrollbars.kt:
```diff
@@ -0,0 +1,218 @@
+private const val INACTIVE_TO_DORMANT_COOL_DOWN = 2_000L
+
+/**
+ * A [Scrollbar] that allows for fast scrolling of content.
```
What does fast mean exactly in this context? How is it different then the other scrollbar?

tunjid replied:
A `FastScrollBar` is one that can be interacted with, it can be dragged. A decorative scrollbar just shows your position in the list.

mmoczkowski commented on core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/component/scrollbar/LazyScrollbarUtilities.kt:
```diff
@@ -0,0 +1,154 @@
+    if (firstItemIndex < 0) return Float.NaN
+
+    val itemOffset = offset(firstItem).toFloat()
+    val offsetPercentage = abs(itemOffset) / itemSize(firstItem)
```
Should we add a sanity check in case `itemSize(firstItem)` returns 0?

tunjid replied:
Great catch, done!

mmoczkowski reviewed (changes requested):
Functional QA results
Devices: Pixel 6 API 32, Pixel 7 API 33, Pixel C API 30

### 1. Scrollbars appear on app launch
- Reproducibility: 100%
- Steps to reproduce:
    1. Open app
- Expected result:
    1. Scrollbars remain hidden 
- Actual result:
    1. Scrollbars appear on app launch and until they fade out 

### 2. Scrollbars appear everytime nav destination is changed
- Reproducibility: 100%
- Steps to reproduce:
    1. Open app
    2. Change navigation tab
- Expected result:
    1. Scrollbars remain hidden 
- Actual result:
    1. Scrollbars appear everytime screen is changed
<img src="https://github.com/android/nowinandroid/assets/6339497/23c15c97-0af9-4540-81fd-202d3aabd690" width="300px"/>

### 3. Scrollbar appears when the container is empty and not scrollable
- Reproducibility: 100%
- Steps to reproduce:
    1. Clear app data
    2. Open app
- Expected result:
    1. Scrollbar remains hidden when list is empty and scrolling is not possible
- Actual result:
    1. Scrollbar appears when list is empty and scrolling is not possible
<img src="https://github.com/android/nowinandroid/assets/6339497/0bddc43d-4f82-411b-88a9-88918b1e80c5" width="300px"/>

### 4. Scrollbar appears when using "swipe left" system gesture
- Reproducibility: 100%
- Steps to reproduce:
    1. Open app
- Expected result:
    1. Scrollbar remains hidden when system gesture is performed
- Actual result:
    1. Scrollbar appears when system gesture is performed
<img src="https://github.com/android/nowinandroid/assets/6339497/f11c36db-f6c0-45b5-9633-bf2cfb303383" width="300px"/>

### 5. Scrolling is not smooth and thumb gets resized
- Reproducibility: 100%
- Steps to reproduce:
    1. Open app
    2. Select few favourite topics
    3. Open "Interests" tab
    4. Scroll the list from top to bottom and reverse
- Expected result:
    1. Scrollbar thumb movement is smooth
    2. Scrollbar thumb does not resize
- Actual result:
    1. Scrollbar thumb movement is not smooth
    2. Scrollbar size keeps resizing when close to edge
<img src="https://github.com/android/nowinandroid/assets/6339497/b003cfd4-0580-4c23-8700-96cd2048bf01" width="300px"/>
<img src="https://github.com/android/nowinandroid/assets/6339497/4994d659-0274-4288-94a4-7c3a199c9ab0" width="300px"/>
<img src="https://github.com/android/nowinandroid/assets/6339497/3aaa699a-f8c4-4c34-acb3-67f97b5d07a6" width="300px"/>
<img src="https://github.com/android/nowinandroid/assets/6339497/8bdf03ec-4cf4-4c0a-b5e9-ce71f75a0f0c" width="300px"/>

### 6. Dragging scrollbar does not result in list being scrolled smoothly
- Reproducibility: 100%
- Steps to reproduce:
    1. Open app
    2. Select few favourite topics
    3. Open "Interests" tab
    4. Scroll the list y dragging scroll thumb from top to bottom and reverse
- Expected result:
    1. List scroll movement is smooth and continous
- Actual result:
    1. List scroll movement is rough and discrete
<img src="https://github.com/android/nowinandroid/assets/6339497/13319491-57c9-4e49-ba3f-1f767aa8e38c" width="300px"/>

### 7. Scrollbars are missing in the search screen
- Reproducibility: 100%
- Steps to reproduce:
    1. Open app
    2. Tap search icon in the top app bar
    3. Search for a topic (i.e. "auto")
- Expected result:
    1. Search screen has scrollbars available
- Actual result:
    1. Search screen has no scrollbars available
<img src="https://github.com/android/nowinandroid/assets/6339497/528eb5e7-1ebc-4241-a19a-32d8fbd0e27a" width="300px"/>

tunjid commented:
@mmoczkowski 

1. Scrollbars appear on app launch
Fixed.

2. Scrollbars appear every time nav destination is changed
Fixed.

3. Scrollbar appears when the container is empty and not scrollable
Fixed.

4. Scrollbar appears when using "swipe left" system gesture
I am fine with this. The alternative would be to use the [systemGestureExclusion](https://developer.android.com/reference/kotlin/androidx/compose/foundation/package-summary#(androidx.compose.ui.Modifier).systemGestureExclusion()) API. However that would disable the back gesture
on screens where the scrollbar is present. What would be helpful however is being able to detect the back gesture and hide the scrollbar if it has been triggered. AFAIK, no such API exists atm.

5. Scrolling is not smooth and thumb gets resized
This unfortunately cannot be avoided and happens with a RecyclerView too.
Since this is a lazy component, the size of each item cannot be fully known and only a
best guess estimate can be done. It helps if all items are fixed in size then the guess can be near exact. Unfortunately in a lazy grid, this is currently impossible bc of this [bug](https://buganizer.corp.google.com/issues/281849377).
   
6. Dragging scrollbar does not result in list being scrolled smoothly
This is also currently impossible. The `scrollToItem` API for lazy columns is not smooth. While `animateScrollToItem` exists, it is not feasible to use with a fast scroller as recomposition causes the animation to be repeatedly cancelled.

7. Scrollbars are missing in the search screen
Fixed.

alexvanyo commented:
For `4`, the platform should enforce a maximum amount of space that can be excluded ([documented as `200dp` here](https://developer.android.com/reference/android/view/View.html#setSystemGestureExclusionRects(java.util.List%3Candroid.graphics.Rect%3E))) so it wouldn't entirely disable the back gesture.

For a really small search bar, that might be nice to use to avoid the conflict.

tunjid commented:
@alexvanyo ah the issue is the scrollbar track touch target. It spans the entire length of the scrollable view, so any gesture along that edge shows the scrollbar to make it discoverable to users.

If I understand @mmoczkowski correctly, the discoverability should only be triggered if the gesture is not eventually resolved to a system back gesture.

mmoczkowski commented:
> If I understand @mmoczkowski correctly, the discoverability should only be triggered if the gesture is not eventually resolved to a system back gesture.

That is correct.

tunjid commented on feature/foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:
```diff
@@ -442,6 +484,26 @@ private fun DeepLinkEffect(
+    onboardingUiState: OnboardingUiState,
+): Int {
+    val feedSize = when (feedState) {
+        NewsFeedUiState.Loading -> 1
```
The ones with size 1 are actually rendered, the ones with size 0 emit no UI.

dturner replied:
Please add a comment. This was (and still is) a [WTF](https://www.osnews.com/story/19266/wtfsm/) for me.

tunjid commented on core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/component/scrollbar/Scrollbar.kt:
```diff
@@ -0,0 +1,353 @@
+ * @param minThumbSize the minimum size of the scrollbar thumb
+ * @param interactionSource allows for observing the state of the scroll bar
+ * @param thumb a composable for drawing the scrollbar thumb
+ * @param onThumbDisplaced an function for reacting to scroll bar displacements caused by direct
```
Technical force of habit; displacement being a vector communicates magnitude and direction, moved like the scalar distance just communicates magnitude. I'm amenable to changing it if it reads easier.

If I did, should all usages of displacement in context change as well?

tunjid commented:
@mmoczkowski the scrollbar appearing when performing the back gesture is fixed. I used modifier.pointerInput and detected horizontal or vertical drags instead of modifier.draggable
